### PR TITLE
Raft group reconfiguration

### PR DESCRIPTION
**Pull request by kbatuigas**
[Deploy preview](https://deploy-preview-262--redpanda-docs-preview.netlify.app/current/manage/raft-group-reconfiguration/)

Adapts the https://redpanda.com/blog/raft-protocol-reconfiguration-solution blog post to our docs. Gives a short introduction on Raft and how it is used in Redpanda, and explains how Redpanda implements group reconfiguration.

Closes https://github.com/redpanda-data/documentation-private/issues/1225.

### CODE REVIEW

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 1]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration in Redpanda
+:description: Learn how the Redpanda Raft group protocol provides consistency and availability during reconfiguration.
+
+Redpanda and Kafka with Raft both use the Raft consensus algorithm to manage cluster metadata. This metadata is stored as messages in an internal topic in a separate Raft cluster instance, called a Raft group. The messages in this topic contain the current set of brokers that are members of a Raft group. In a Raft group, one member is the leader and the others are followers.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: we name the special metadata raft group a controller. each raft instance both data and metadata forms a raft group.

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+== About Raft group reconfiguration
+
+A Raft group’s member set changes during operations like cluster resize or xref:manage:cluster-maintenance/continuous-data-balancing.adoc[Continuous Data Balancing]. This change, also referred to as Raft group reconfiguration, includes adding, removing, or replacing Raft group members. When the group configuration changes, the current leader replicates a new configuration message and sends it to the followers. 
+
+Raft uses the joint consensus approach to maintain consistency during reconfiguration. It involves both the old member set and the proposed new member set electing a new group leader and agreeing on the latest committed offset. The new leader can be elected from either the old or new member set. Partitions are unavailable to accept writes until the group reaches joint consensus.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: the last part i.e. group reaches joint consensus, i would rephrase as joint consensus is a name of reconfiguration algorithm. Here both old and new replica set must agree to accept a write or vote for a leader

File: modules/manage/pages/raft-group-reconfiguration.adoc (hunk 1 above)
Comment (kbatuigas) [reply modules/manage/pages/raft-group-reconfiguration.adoc]: @mmaslankaprv to double check - is that a different `controller` than the one defined here? https://docs.redpanda.com/current/reference/glossary/#controller

Comment (mmaslankaprv) [reply modules/manage/pages/raft-group-reconfiguration.adoc]: We have a special partition (partition == raft group i.e. raft + storage) called `redpanda/controller/0` this raft group stores all the metadata that are mentioned. A leader of this Raft group is a controller broker. 
A raft group term is a general term describing single Raft protocol instance

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+== About Raft group reconfiguration
+
+A Raft group’s member set changes during operations like cluster resize or xref:manage:cluster-maintenance/continuous-data-balancing.adoc[Continuous Data Balancing]. This change, also referred to as Raft group reconfiguration, includes adding, removing, or replacing Raft group members. When the group configuration changes, the current leader replicates a new configuration message and sends it to the followers. 
+
+Raft uses the joint consensus algorithm to maintain consistency during reconfiguration. It involves both the old member set and the proposed new member set reaching an agreement between them, either on accepting writes, or electing a new leader. The new leader can be elected from either the old or new member set. Partitions are unavailable to accept writes until the old and new sets reach joint consensus.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: please remove that last sentence: 
```
Partitions are unavailable to accept writes until the old and new sets reach joint consensus.
```
Partitions may accept writes all the time during reconfiguration.

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+Redpanda’s implementation of Raft includes changes to the reconfiguration protocol to ensure high availability even while group membership changes.
+
+== Redpanda’s approach to Raft reconfiguration
+
+Redpanda’s Raft implementation adds the following behavior when reconfiguring Raft groups:
+
+- In addition to leader and follower, a broker can also be in learner status. Learners are followers that cannot vote for a new leader, so they do not count towards quorum. Learners do not trigger leader elections and do not accept writes. New Raft group members start as learners. Brokers can be promoted or demoted between learner and voter.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: learners can vote, they can not become a leader, and they are not accounted into majority

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
++
+This means that when new members are added to the Raft group member set, the group continues to operate without interruption since members that can form quorum still accept writes, even if the majority has technically changed due to the additional members. As soon as a learner is synced with the current leader, the learner is promoted to voter. Learners are promoted one at a time. 
+
+- Because all new nodes joining a Raft group start in learner status, the group does not yet have to enter the joint consensus phase as soon as new members join.
+
+- The Raft group only adds new members first. When all learners have been promoted to voters, the group can enter joint consensus. At this point, this is also a new group configuration that contains both the old as well as requested new member set. The current leader replicates this configuration to all brokers. 
++
+Only after learners are up to date with the leader will the group elect a new leader through joint consensus. Members that are not needed in the new configuration are demoted to learners before removal. The new leader then replicates the desired final configuration.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: This `Partitions are unavailable to accept writes until the old and new sets reach joint consensus.` can be removed

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 6]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration in Redpanda
+:description: Learn how the Redpanda Raft group protocol provides consistency and availability during reconfiguration.
+
+Redpanda and Kafka with Raft (KRaft) both use the https://raft.github.io/[Raft consensus algorithm] to manage cluster metadata.Redpanda stores this metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group. 
+
+Unlike KRaft, Redpanda also uses Raft as its replication protocol for the actual data (that is, the application data). Each topic partition in Redpanda forms a Raft group comprised of its leader and follower brokers. Redpanda simplifies Kafka architecture because Raft handles both metadata management as well as data replication. See xref:get-started:architecture.adoc#raft-consensus-algorithm[How Redpanda Works] for a further explanation of how Raft manages data replication in Redpanda.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Unlike Kafka with Raft (KRaft), Redpanda also uses Raft as its replication protocol for the application data. Each topic glossterm:partition[] in Redpanda forms a Raft group comprised of its leader and follower brokers. Redpanda simplifies Kafka architecture because Raft handles both metadata management as well as data replication. See xref:get-started:architecture.adoc#raft-consensus-algorithm[How Redpanda Works] for a further explanation of how Raft manages data replication in Redpanda.
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 7]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration in Redpanda
+:description: Learn how the Redpanda Raft group protocol provides consistency and availability during reconfiguration.
+
+Redpanda and Kafka with Raft (KRaft) both use the https://raft.github.io/[Raft consensus algorithm] to manage cluster metadata.Redpanda stores this metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Redpanda uses the glossterm:Raft[] consensus algorithm to manage cluster metadata. Redpanda stores this metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group. 
```

Should we update our glossary definition to `controller broker`? https://raw.githubusercontent.com/redpanda-data/docs/shared/modules/terms/partials/controller.adoc

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+== About Raft group reconfiguration
+
+A Raft group’s member set changes during operations like cluster resize or xref:manage:cluster-maintenance/continuous-data-balancing.adoc[Continuous Data Balancing]. This change, also referred to as Raft group reconfiguration, includes adding, removing, or replacing Raft group members. When the group configuration changes, the current leader replicates a new configuration message and sends it to the followers. 
+
+Raft uses joint consensus to maintain consistency during reconfiguration. Joint consensus involves both the old member set and the proposed new member set reaching an agreement between them, either on accepting writes, or electing a new leader. The new leader can be elected from either the old or new member set.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Raft uses joint consensus to maintain consistency during reconfiguration. Joint consensus involves both the old member set and the proposed new member set reaching an agreement, either on accepting writes or electing a new leader. The new leader can be elected from either the old or new member set. 
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+== Redpanda’s approach to Raft reconfiguration
+
+Redpanda’s Raft implementation adds the following behavior when reconfiguring Raft groups:
+
+- In addition to leader and follower, a broker can also be in learner status. Learners are followers that do not count towards quorum (the majority) and cannot be elected to leader. Learners do not trigger leader elections and do not accept writes. New Raft group members start as learners. Brokers can be promoted or demoted between learner and voter. 
++
+This means that when new members join the Raft group member set, the group continues to operate without interruption since members that can form quorum still accept writes, even if the majority has technically changed due to the additional members. As soon as a learner is synced with the current leader, the learner is promoted to voter. Learners are promoted one at a time. 
+
+- Because all new nodes joining a Raft group start in learner status, the group does not yet have to enter the joint consensus phase as soon as new members join.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]: Check for instances of nodes vs brokers.

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+- The Raft group only adds new members first. When all learners have been promoted to voters, the group can enter joint consensus. At this point, this is also a new group configuration that contains both the old as well as requested new member set. The current leader replicates this configuration to all brokers. 
++
+Only after learners are up to date with the leader will the group elect a new leader through joint consensus. Members that are not in the new configuration are demoted to learners before removal. The new leader then replicates the desired final configuration.
+
+The following diagrams illustrate the reconfiguration steps in Redpanda. In this example, we replace one member in a three-broker member set of a Raft group. The box labeled "new" represents the new or current configuration, and "old" represents the old or previous configuration. Each configuration includes the member brokers as indicated by their IDs. Brokers in green are voters, whereas brokers in yellow are learners. If the old configuration is grayed out, it means that only the new configuration is active.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
The following diagrams illustrate the reconfiguration steps in Redpanda. In this example, one member joins the three-broker member set of a Raft group. The box labeled "new" represents the new or current configuration, and "old" represents the old or previous configuration. Each configuration includes the member brokers as indicated by their IDs. Brokers in green are voters, whereas brokers in yellow are learners. If the old configuration is grayed out, it means that only the new configuration is active.
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+
+. We start with a configuration containing brokers 1, 2, and 3:
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
. Start with a configuration containing brokers 1, 2, and 3:
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
++
+image::shared:raft_reconfiguration_initial_state.png[]
+
+. We want to replace broker 1 with a new member broker 4. Broker 4 is added to the current configuration as a learner:
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
. Replace broker 1 with a new member (broker 4). Broker 4 is added to the current configuration as a learner:
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+. Broker 1 is demoted to learner:
++
+image::shared:raft_reconfiguration_voter_demote_to_learner.png[]
+
+. Broker 1 is removed. The leader replicates the new configuration:
++
+image::shared:raft_reconfiguration_final_state.png[]
+
+A Raft group that can tolerate _F_ broker failures needs to have at least 2__F__ + 1 members. The number of members is typically equal to the replication factor. In our example, the group can tolerate single-broker failure. The two remaining healthy group members can still reach quorum.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
A Raft group that can tolerate _F_ broker failures needs to have at least 2__F__ + 1 members. The number of members is typically equal to the glossterm:replication factor[]. In this example, the group can tolerate a single-broker failure. The two remaining healthy group members can still reach quorum.
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 14]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration in Redpanda
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
= Raft Group Reconfiguration
```

**Review by JakeSCahill — COMMENTED** [review]
Looking good. Left some suggestions mainly to use our glossary more for this. Consider adding more terms such as learners to the glossary.

File: modules/manage/pages/raft-group-reconfiguration.adoc (hunk 7 above)
Comment (kbatuigas) [reply modules/manage/pages/raft-group-reconfiguration.adoc]: @JakeSCahill I think it is a bit confusing that `controller` in this context can mean a "special metadata Raft group" and then the broker that is the current leader of the group is `controller broker`, but in general, `controller` refers to a single broker, and we'd just say `partition leader` (glossary term [here](https://docs.redpanda.com/current/reference/glossary/#partition-leader)) to refer to the Raft group leader. 

@mmaslankaprv is it possible to resolve this? Should we change `controller` to `controller broker` in the glossary? Or would it be better to remove `A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker.` entirely? Perhaps these are internal details that don't need to be spelled out to readers?

Comment (mmaslankaprv) [reply modules/manage/pages/raft-group-reconfiguration.adoc]: I think the `controller broker` will prevent confusion.

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+Raft uses joint consensus to maintain consistency during reconfiguration. Joint consensus involves both the old member set and the proposed new member set reaching an agreement, either on accepting writes or electing a new leader. The new leader can be elected from either the old or new member set. 
+
+Redpanda’s implementation of Raft includes changes to the reconfiguration protocol to ensure high availability even as group membership changes.
+
+== Redpanda’s approach to Raft reconfiguration
+
+Redpanda’s Raft implementation adds the following behavior when reconfiguring Raft groups:
+
+- In addition to leader and follower, a broker can also be in learner status. Learners are followers that do not count towards quorum (the majority) and cannot be elected to leader. Learners do not trigger leader elections and do not accept writes. New Raft group members start as learners. Brokers can be promoted or demoted between learner and voter.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
- In addition to leader and follower, a broker can also be in learner status. Learners are followers that do not count towards quorum (the majority) and cannot be elected to leader. Learners do not trigger leader elections. New Raft group members start as learners. Brokers can be promoted or demoted between learner and voter. 
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
++
+This means that when new members join the Raft group member set, the group continues to operate without interruption since members that can form quorum still accept writes, even if the majority has technically changed due to the additional members. As soon as a learner is synced with the current leader, the learner is promoted to voter. Learners are promoted one at a time. 
+
+- Because all new brokers joining a Raft group start in learner status, the group does not yet have to enter the joint consensus phase as soon as new members join.
+
+- The Raft group only adds new members first. When all learners have been promoted to voters, the group can enter joint consensus. At this point, this is also a new group configuration that contains both the old as well as requested new member set. The current leader replicates this configuration to all brokers. 
++
+Only after learners are up to date with the leader will the group elect a new leader through joint consensus. Members that are not in the new configuration are demoted to learners before removal. The new leader then replicates the desired final configuration.
```
Comment (mmaslankaprv) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Members that are not in the new configuration are demoted to learners before removal.  When Raft group leaves joint consensus mode a new leader may be elected if the old one was removed from configuration. 
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 17]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration
+:description: Learn how the Redpanda Raft group protocol provides consistency and availability during reconfiguration.
+
+Redpanda uses the glossterm:Raft[] consensus algorithm to manage cluster metadata. Redpanda stores this metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Raft group reconfiguration in Redpanda involves dynamically changing the membership of a Raft group to ensure consistent and available distributed system operations. This process uses learners, new members that do not affect quorum, to maintain service continuity during additions, removals, or replacements in the group.

== How Redpanda uses Raft

Redpanda uses the glossterm:Raft[] consensus algorithm to replicate cluster metadata and application data across the brokers. This replication provides strong guarantees for data safety and fault tolerance.

Redpanda stores cluster metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group. 
```

File: modules/manage/pages/raft-group-reconfiguration.adoc [hunk 18]
```diff
@@ -0,0 +1,66 @@
+= Raft Group Reconfiguration
+:description: Learn how the Redpanda Raft group protocol provides consistency and availability during reconfiguration.
+
+Redpanda uses the glossterm:Raft[] consensus algorithm to manage cluster metadata. Redpanda stores this metadata as messages in a special topic in a Raft cluster instance, also known as a Raft group. In a Raft group, one member is the leader and the others are followers. A Raft group that stores cluster metadata is called a controller, and the leader of this group is the controller broker. The messages in the metadata topic contain the current set of brokers that are members of the group. 
+
+Unlike Kafka with Raft (KRaft), Redpanda also uses Raft as its replication protocol for the application data. Each topic glossterm:partition[] in Redpanda forms a Raft group comprised of its leader and follower brokers. Redpanda simplifies Kafka architecture because Raft handles both metadata management as well as data replication. See xref:get-started:architecture.adoc#raft-consensus-algorithm[How Redpanda Works] for a further explanation of how Raft manages data replication in Redpanda.
```
Comment (JakeSCahill) [inline modules/manage/pages/raft-group-reconfiguration.adoc]:
```suggestion
Unlike Kafka with Raft (KRaft), Redpanda uses Raft not only for cluster metadata but also for the application data. Each topic glossterm:partition[] in Redpanda forms a Raft group comprised of its leader and follower brokers. Redpanda simplifies Kafka architecture because Raft handles both metadata management as well as data replication. See xref:get-started:architecture.adoc#raft-consensus-algorithm[How Redpanda Works] for a further explanation of how Raft manages data replication in Redpanda.
```

File: modules/manage/pages/raft-group-reconfiguration.adoc (hunk 17 above)
Comment (JakeSCahill) [reply modules/manage/pages/raft-group-reconfiguration.adoc]: The first couple of paragraphs didn't cover the title of the topic. As a reader, I'd like to learn what Raft group reconfiguration is and why it's important in the first paragraph. Feel free to rephrase though 😄

**Review by JakeSCahill — APPROVED** [review]
Other than the suggestion for a new opening paragraph LGTM!

**Comment by kbatuigas** [comment]
Hi @mmaslankaprv , since this PR has now been reviewed multiple times and approved by the docs team, we'll go ahead and  merge this. If you spot anything that needs to be edited, please let us know, or feel free to open a new PR. Thank you for your help with this!
